### PR TITLE
Move try-except KeyboardInterrupt around get() setup

### LINKED ISSUE #1439: Dask doesn't play well with interrupts
This is on `Python 3.5.2`, `dask 0.10.1`, and `jupyter-notebook 4.2.1`, all installed via conda on a 64-bit Ubuntu machine.

If you run the following piece of code:
```Python
import json

from dask.diagnostics import ProgressBar
import dask.bag as db

j = json.dumps({"a": 1, "b": 1})

for i in range(8):
    data = [j for _ in range(10 ** i)]
    bag = db.from_sequence(data, npartitions=4).map(json.loads)
    with ProgressBar():
        db.zip(bag.pluck("a"), bag.pluck('b')).count().compute()
```

And try to do a keyboard interrupt in the middle, sometimes dask will flip out and just refuse to exit (no matter how many `CTRL-C`s and `CTRL-D`s you press. See https://asciinema.org/a/2yqxbhn1patwenwy024m2316l for a video.

This is especially annoying on a Jupyter notebook. Sometimes, dask will mysteriously keep printing out the progress bar from the last session, even after restarting the kernel (!).

_EDIT_: When this happens, dask will also leave around `python` processes still running. See https://asciinema.org/a/8x1liiyuiw7910mugj9tscher for an example (particularly the extra `python` process at the end).

### PR DESCRIPTION
**Pull request by mrocklin**
Previously when calling dask.async.get_async we would only catch
KeyboardInterrupts that happened during computation, not during setup
and teardown.  This could cause an issue if a ProgressBar callback
was setup but then the KeyboardInterrupt happened either during further
setup or after computation during teardown (but before ProgressBar
teardown)

This commit expands the range of the KeyboardInterrupt try-except to
include everything after plugin registration.

Fixes #1439

### CODE REVIEW

**Comment by mrocklin** [comment]
cc @jcrist for review

Thanks to @alanhdu and @rolando for reporting in #1439

File: dask/async.py [hunk 1]
```diff
@@ -426,69 +426,69 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
...
 
-    dsk = dsk.copy()
-    for f in start_cbs:
-        f(dsk)
-
-    dsk, dependencies = cull(dsk, list(results))
+    try:
+        dsk = dsk.copy()
+        for f in start_cbs:
+            f(dsk)
```
Comment (jcrist) [inline dask/async.py]: This opens the (small) chance that a callback will have `finish` called without having `start` ever called. The existing callbacks may need to be fixed to be robust against this case.

Comment (mrocklin) [inline dask/async.py]: Small chances were what got us into this mess (the computation in the issue was a very fast and repeated one to spend more time in these sections).  We could also make a `started_cbs` list.

Comment (jcrist) [inline dask/async.py]: > We could also make a started_cbs list.

I like that option better - keeps the callback code simpler. Good idea.

Comment (mrocklin) [inline dask/async.py]: Hrm, this seems to interfere with the `unpack_callbacks` bit.  `started_cbs` and `finish_cbs` don't necessarily align.

Do we know how expensive the extra function calls are that `unpack_callbacks` saves us from?

Comment (jcrist) [inline dask/async.py]: Should be negligible for start/finish stuff. I'd expect slightly less negligible for pre/post task, but it probably doesn't matter there either.

Comment (mrocklin) [inline dask/async.py]: Fixed.  Review appreciated.  It looks a bit uglier now.

Comment (jcrist) [inline dask/async.py]: Looks good enough to me, and doesn't require too much extra logic. +1.

**Pull request merged**